pyo3-macros-backend/src/pyclass.rs

```diff
@@ -178,7 +194,16 @@ impl PyClassPyO3Options {
             PyClassPyO3Option::SetAll(set_all) => set_option!(set_all),
             PyClassPyO3Option::Subclass(subclass) => set_option!(subclass),
             PyClassPyO3Option::Unsendable(unsendable) => set_option!(unsendable),
-            PyClassPyO3Option::Weakref(weakref) => set_option!(weakref),
+            PyClassPyO3Option::Weakref(weakref) => {
+                if python_version >= PY_3_9 || !is_abi3() {
+                    set_option!(weakref);
+                } else {
+                    return Err(syn::Error::new(
+                        weakref.span(),
+                        "`weakref` requires >= python 3.9 with the `abi3` feature",
+                    ));
+                }
```

Comment:
Same suggestion to use `ensure_spanned!`

```suggestion
                ensure_spanned!(
                    python_version >= PY_3_9 || !is_abi3(),
                    weakref.span() => "`weakref` requires >= python 3.9 with the `abi3` feature"
                );
                set_option!(weakref);
```